mysql-test/mysql-test-run.pl

```diff
@@ -4003,8 +4003,8 @@ ($$)
       else
       {
          $proc= My::SafeProcess->wait_any_timeout($test_timeout);
-         $keep_waiting_proc{$proc} = 1;
       }
+      $keep_waiting_proc{$proc} = 1;
     }
 
     if (scalar(keys(%keep_waiting_proc)) == 0)
```

Comment:
this changes behavior. Old code (before your first patch) did "wait_any failed" error only if $proc was empty, when wait_any() failed. Now you get in here also when
```
	# Also check if timer has expired, if so cancel waiting
	if ( has_expired($test_timeout) )
	{
	  %keep_waiting_proc = ();
	}
```

I think this should be fixed.

Reply:
I think the "wait_any failed" piece was a dead code before my changes:

```
3970     my $proc;
3971     if ($keep_waiting_proc)
3972     { 
3973       # Any other process exited?
3974       $proc = My::SafeProcess->check_any();
3975       if ($proc)
3976       { 
3977         mtr_verbose ("Found exited process $proc");
3978       }
3979       else
3980       { 
3981         $proc = $keep_waiting_proc;
3982         # Also check if timer has expired, if so cancel waiting
3983         if ( has_expired($test_timeout) )
3984         { 
3985           $keep_waiting_proc = 0;
3986         }
3987       }
3988     }
3989     if (! $keep_waiting_proc)
3990     { 
3991       if($test_timeout > $print_timeout)
3992       {  
3993          $proc= My::SafeProcess->wait_any_timeout($print_timeout);
3994          if ( $proc->{timeout} )
3995          {  
3996             #print out that the test is still on
3997             mtr_print("Test still running: $tinfo->{name}");
3998             #reset the timer
3999             $print_timeout= start_timer($print_freq * 60);
4000             next;
4001          }
4002       }
4003       else
4004       {
4005          $proc= My::SafeProcess->wait_any_timeout($test_timeout);
4006       }
4007     }
4008 
4009     # Will be restored if we need to keep waiting
4010     $keep_waiting_proc = 0;
4011 
4012     unless ( defined $proc )
4013     {
4014       mtr_error("wait_any failed");
4015     }
```

`$proc` would always be defined on line 4012 and 4014 would never execute. Let me elaborate:

If `3971     if ($keep_waiting_proc)` evaluates to true, then `$proc` will be assigned from `My::SafeProcess->check_any();` and if that returns `undefined`, then on line 3981 `$proc` is assigned from `$keep_waiting_proc` which we know is defined (because the condition on line 3971 was true).

If `3971     if ($keep_waiting_proc)` evaluates to false, then `3989     if (! $keep_waiting_proc)` will evaluate to true. So, on line 3971 we either have `$proc` defined (because 3971 was true) or the condition on 3971 will evaluate to true. If the latter happens, then inside the body of the second if (lines 3991-4006) `$proc` will get defined because it is assigned from `My::SafeProcess->wait_any_timeout()` which always returns an object (never `undef` or 0).

So, line 4014 will never execute. I have removed this code.